### PR TITLE
Auth 42 api

### PR DESCRIPTION
**Pull request by tsiguenz**

### CODE REVIEW

File: backend/prisma/schema.prisma [hunk 1]
```diff
@@ -17,12 +17,13 @@ enum GameType {
...
 model User {
   id                  Int       @default(autoincrement()) @id
+  fortyTwoId          Int?      @unique
   nickname            String    @unique
   ladderPoints        Int       @default(0)
   avatar              Bytes? 
   twoFactorEnable     Boolean   @default(false)
   twoFactorSecret     String?
-  hash                String
+  hash                String?
```
Comment (Crysicia) [inline backend/prisma/schema.prisma:26]: Can't we signin w/o password then ?

File: backend/src/users/users.service.ts [hunk 2]
```diff
@@ -2,6 +2,7 @@ import { Injectable, NotFoundException } from '@nestjs/common';
 import { PrismaService } from '../prisma/prisma.service';
 import { JwtService } from '@nestjs/jwt';
 import { UnauthorizedException, ForbiddenException } from '@nestjs/common';
+import * as os from 'os';
```
Comment (Crysicia) [inline backend/src/users/users.service.ts:5]: Why?

Comment (Crysicia) [inline backend/src/users/users.service.ts:5]: Delete

File: todo.txt [hunk 3]
```diff
@@ -1,6 +0,0 @@
-- define the api use-case and implement
-- prepare the database diagram
-- generate the 42api and test it
-- connect backend to database
-- connect frontend to backend
-- configure vite
```
Comment (Crysicia) [inline todo.txt:6]: Gotta go fast
Gotta go fast

**Review by Crysicia — CHANGES_REQUESTED** [review]

File: todo.txt (hunk 3 above)
Comment (tsiguenz) [reply todo.txt:6]: Don't judge my notes pls :cry:

File: backend/prisma/schema.prisma (hunk 1 above)
Comment (tsiguenz) [reply backend/prisma/schema.prisma:26]: It's fix

File: backend/src/auth/auth.controller.ts [hunk 4]
```diff
@@ -48,4 +61,47 @@ export class AuthController {
   signin(@Body() dto: AuthDto) {
     return this.authService.signin(dto);
   }
+
+  @UseGuards(FortyTwoGuard)
+  @Get('42')
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  async fortyTwoAuth() {}
```
Comment (Crysicia) [inline backend/src/auth/auth.controller.ts:68]: Why is this needed?
Is it the callback URL?

File: backend/src/auth/auth.service.ts [hunk 5]
```diff
@@ -46,24 +51,20 @@ export class AuthService {
...
-    }
+    if (!user || user.fortyTwoId)
+      throw new ForbiddenException('Invalid password');
     const valid = await argon.verify(user.hash, dto.password);
     if (!valid) {
       throw new ForbiddenException('Invalid password');
     }
     if (user.twoFactorEnable) {
-      if (!dto.twoFactorCode) {
+      if (!dto.twoFactorCode)
```
Comment (Crysicia) [inline backend/src/auth/auth.service.ts:61]: Could be one line.

File: backend/src/auth/auth.service.ts [hunk 6]
```diff
@@ -74,14 +75,98 @@ export class AuthService {
...
+    if (user.twoFactorEnable) {
+      res.send(
+        `<form action="/api/auth/42?state=${state}" method="post"> \
+        <label for="twoFa">Two factor code :</label> \
+        <input id="twoFa" name="twoFa" autofocus></input> \
+        <button type="submit">Submit</button> \
+        </form>`
+      );
+      return;
+    }
```
Comment (Crysicia) [inline backend/src/auth/auth.service.ts:126]: What the F is this?!
Can't we just redirect the user to a 2FA form?

File: backend/src/auth/auth.service.ts [hunk 7]
```diff
@@ -74,14 +75,98 @@ export class AuthService {
...
+    res.cookie('jwt', jwt.access_token);
+    return res.redirect(`http://${process.env.HOST_IP}:8080/home`);
+  }
+
+  redirectionAlert(res: Response, message: string, url: string) {
+    const send = ` <script type="text/javascript"> \
+            alert("${message}"); \
+            window.location.href = "${url}"; \
+          </script>`;
+    res.send(send);
```
Comment (Crysicia) [inline backend/src/auth/auth.service.ts:137]: Same here.

File: backend/src/auth/auth.service.ts [hunk 8]
```diff
@@ -74,14 +75,98 @@ export class AuthService {
...
+      twoFaCode.toString()
+    );
+    if (!valid) {
+      this.redirectionAlert(res, 'Invalid two factor code', urlSignin);
+      return;
+    }
+    const jwt = await this.createJwt(user.id);
+    res.cookie('jwt', jwt.access_token);
+    return res.redirect(`http://${process.env.HOST_IP}:8080/home`);
+  }
```
Comment (Crysicia) [inline backend/src/auth/auth.service.ts:171]: Do we really need to handle 2FA with Oauth separately from 2FA with Uaer/Password?
Seems like code duplication.

File: frontend/src/views/Signin.vue [hunk 9]
```diff
@@ -32,6 +32,10 @@
     ></v-text-field>
     <v-btn @click="signin">Sign In</v-btn>
   </v-form>
+  <br />
```
Comment (Crysicia) [inline frontend/src/views/Signin.vue:35]: We hate line breaks.
Maybe a v-card could do the trick.

**Review by Crysicia — APPROVED** [review]
All in all it's good.
Thanks for your service, you did a great job!

File: frontend/src/views/Signin.vue (hunk 9 above)
Comment (tsiguenz) [reply frontend/src/views/Signin.vue:35]: I'm not the frontend guy

File: backend/src/auth/auth.service.ts (hunk 6 above)
Comment (tsiguenz) [reply backend/src/auth/auth.service.ts:126]: You just doesn't like my art :cry:

File: backend/src/auth/auth.controller.ts (hunk 4 above)
Comment (tsiguenz) [reply backend/src/auth/auth.controller.ts:68]: No, it's empty function to call the 42 strategy who redirect to the callback.
It's ugly but everyone do this like that...

File: backend/src/auth/auth.service.ts (hunk 6 above)
Comment (tsiguenz) [reply backend/src/auth/auth.service.ts:126]: Yes we can ! Thanks
